libnetwork/d/o/ovmanager: switch to using bitmap package directly

corhere opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Substituted `idm.Idm` with `bitmap.Bitmap`.

**- How I did it**
By cleaning up locking and using the ovmanager driver's mutex to guard against concurrent access to the bitmap.

**- How to verify it**
By visual inspection?

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
libn/d/o/ovmanager: make mutexes private fields

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
libn/d/o/ovmanager: drop mutex from network type

In the network.obtainVxlanID() method, the mutex only guards a local
variable and a function argument. Locking is therefore unnecessary.

The network.releaseVxlanID() method is only called in two contexts:
driver.NetworkAllocate(), where the network struct is a local variable
and network.releaseVxlanID() is only called in failure code-paths in
which the network does not escape; and driver.NetworkFree(), while the
driver mutex is held. Locking is therefore unnecessary.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
libn/d/o/ovmanager: inline obtainVxlanID method

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
libn/d/o/ovmanager: assign vxlans using bitmap pkg

The idm package wraps bitseq.Handle to provide an offset and
synchronization. bitseq.Handle wraps bitmap.Bitmap to provide
persistence in a datastore. As no datastore is passed and the offset is
zero, the idm.Idm instance is nothing more than a concurrency-safe
wrapper around a bitmap.Bitmap with differently-named methods. Switch
over to using bitmap.Bitmap directly, using the ovmanager driver's mutex
for concurrency control.

Hold the driver mutex for the entire duration that VXLANs are being
assigned to the new network. This makes allocating VXLANs for a network
an atomic operation.

Signed-off-by: Cory Snider <csnider@mirantis.com>

neersighted commented on libnetwork/drivers/overlay/ovmanager/ovmanager.go:
```diff
@@ -111,11 +109,18 @@ func (d *driver) NetworkAllocate(id string, option map[string]string, ipV4Data,
+				n.releaseVxlanID()
+				return nil, fmt.Errorf("could not assign vxlan id %v to pool %s: %v", s.vni, s.subnetIP, err)
+			}
+		} else {
```
It was a little tricky to figure out what the existing logic was doing, but for anyone else reviewing:

`s.vni` would have been the zero value due to the earlier `len(vxlanIDList) > i` conditional by the time we hit the old `obtainVxlanID`, so this is a transparent simplification.

cpuguy83 replied:
Maybe lets get a comment here?
This is pretty difficult to understand at a glance.

cpuguy83 replied:
Also to note, I think the logic above this loop (which does the `strings.Split`) is fairly expensive (or at least has potential to be).

cpuguy83 replied:
I honestly don't understand why `vxlanIDList` corresponds to anything in `ipv4Data`, it would be great to have a comment around this... or for that matter why the5re isn't something similar for ipv6 (other than I guess its just not implemented in overlay?)

corhere replied:
> Also to note, I think the logic above this loop (which does the `strings.Split`) is fairly expensive (or at least has potential to be).

The lock is not taken until after that logic.

> I honestly don't understand why `vxlanIDList` corresponds to anything in `ipv4Data`, it would be great to have a comment around this...

I don't understand why a separate VNI is allocated per subnet either. And in true libnetwork fashion, the original rationale for _why_ it was [implemented that way](https://github.com/moby/libnetwork/pull/584) has been lost to the mists of time. All I can do is speculate.

> or for that matter why the5re isn't something similar for ipv6 (other than I guess its just not implemented in overlay?)

- It's not implemented: #43615

cpuguy83 replied:
> The lock is not taken until after that logic.

I'm more referring to the potentially unbounded slice that gets created over using an approach that scans over the string.
It may even be possible to use repeated `strings.Cut` calls in the for loop.

corhere replied:
Fair enough. That's certainly worth addressing, though it's out of scope of this PR. The similar code in the non-manager overlay driver should also be updated at the same time. https://github.com/moby/moby/blob/11c6ec6eb2033600528edcb08c57a88d45dc2e9a/libnetwork/drivers/overlay/ov_network.go#L108-L122

corhere replied:
https://github.com/moby/moby/pull/45892

neersighted reviewed (approved):
LGTM

thaJeztah commented on libnetwork/drivers/overlay/ovmanager/ovmanager.go:
```diff
@@ -160,38 +159,26 @@ func (d *driver) NetworkFree(id string) error {
+	vni := uint64(s.vni)
 	if vni == 0 {
-		vni, err = n.driver.vxlanIdm.GetIDInRange(vxlanIDStart, vxlanIDEnd, true)
+		vni, err := n.driver.vxlanIdm.GetIDInRange(vxlanIDStart, vxlanIDEnd, true)
```
This chunk is slightly confusing, as it looks like we're deliberately creating a new `vni` variable, but with the same name as outside the block. The code is _correct_, but wondering if it would be more clear to either use a different variable name, or perhaps we should reverse the conditions;

```go
func (n *network) obtainVxlanID(s *subnet) error {
	if vni := uint64(s.vni); vni != 0 {
		return n.driver.vxlanIdm.GetSpecificID(vni)
	}

	vni, err := n.driver.vxlanIdm.GetIDInRange(vxlanIDStart, vxlanIDEnd, true)
	if err != nil {
		return err
	}

	s.vni = uint32(vni)
	return nil
}
```

Very much side-note; it's a bit awkward that we have to do the uint64 <--> uint32 shuffle here 😞

thaJeztah replied:
DOH! The next commit removes this chunk, so guess there's no need to update the intermediate commit

thaJeztah commented on libnetwork/drivers/overlay/ovmanager/ovmanager.go:
```diff
@@ -52,22 +52,19 @@ func Init(dc driverapi.DriverCallback, _ map[string]interface{}) error {
 	})
 }
 
+func new() *driver {
```
This collides with the `new` built-in; perhaps we should pick a different name (`newDriver()` ? I see we removed that from the `_test.go`, so we can pick its spot)

(trying to keep the linters at bay 😅)

neersighted replied:
How's `func create()`?

corhere replied:
Whoops, I totally forgot that `new` was a builtin!

thaJeztah reviewed (commented):
LGTM, but left one comments on the `new` function colliding with the builtin 🙈

cpuguy83 commented on libnetwork/drivers/overlay/ovmanager/ovmanager.go:
```diff
@@ -53,22 +52,19 @@ func Init(dc driverapi.DriverCallback, _ map[string]interface{}) error {
+func newDriver() *driver {
+	return &driver{
+		networks: networkTable{},
+		vxlanIdm: bitmap.New(vxlanIDEnd + 1),
```
I know this isn't technically new code but can we get a comment on what the logic is behind this value?

cpuguy83 commented on libnetwork/drivers/overlay/ovmanager/ovmanager.go:
```diff
@@ -159,42 +162,15 @@ func (d *driver) NetworkFree(id string) error {
-
 func (n *network) releaseVxlanID() {
-	n.Lock()
 	vnis := make([]uint32, 0, len(n.subnets))
```
Given that we are holding the lock during the entirety of this function call now, does it make sense to create this intermediate slice?
Seems like this could be pretty expensive.

cpuguy83 replied:
Or is `Release` expensive enough that we shouldn't be holding the lock during this?

corhere replied:
> Given that we are holding the lock during the entirety of this function call now, does it make sense to create this intermediate slice?

No, it does not. Good catch, fixed!

`len(n.subnets)` should be small since the user has to enumerate all the subnets when creating the network, and `Release`/`Unset` shouldn't be too expensive. The bitmap is not safe for concurrent use so has to be accessed within a critical section. And if the bitmap operation does prove to be expensive enough to be a problem, I would solve that by replacing our bespoke bitmap implementation with [roaring](https://pkg.go.dev/github.com/RoaringBitmap/roaring).

corhere committed:
libnetwork/bitmap: improve documentation

Clarify that the argument to New is an exclusive upper bound.

Correct the documentation for SetAnyInRange: the end argument is
inclusive rather than exclusive.

Signed-off-by: Cory Snider <csnider@mirantis.com>
